### ISSUE TITLE
[RFC] Upstream Flash DataPipeline to Lightning

### ISSUE THREAD

**Issue by ethanwharris** [PyTorchLightning/pytorch-lightning#7902]
## 🚀 Feature

This feature would up stream the Flash data pipeline to lightning

### Motivation

Data pipeline offers useful data loading and transformation hooks that are valuable to a wider audience than just Flash

### Pitch

We could upstream the following components:

- `DataPipeline` - Really just a wrapper for the other classes
- `DataSource` - A base class for creating multiple data sets just from some metadata (e.g. a folder name)
- `Preprocess` - Provides hooks for batch and sample level transforms on and off device (pre-model)
- `Postprocess` - Provides hooks for batch and sample level transforms on and off device (post-model)
- `Serializer` - Provides a hook for serializing predictions (e.g. convert logits tensor to labels)
- `DataPipelineState` - enables sharing of some state across the pipeline (e.g. labels inferred in `DataSource` can be shared with the `Serializer` to give predictions as labels
- `DataModule` - an extension of the lightning data module that applies the data pipeline components (would need to be renamed)

These are just the base classes, actual implementations can stay in Flash. Lightning users could implement some or all of the base classes to customize data loading to their needs.

#### Docs

The docs are here: https://lightning-flash.readthedocs.io/en/latest/general/data.html
These would need to be cleaned up during the upstreaming as they are not currently in great shape, but the bones are there.

#### Tests

We do have tests already, but these will likely need a review and some expansion.

### Alternatives

Don't have these objects in lightning.

**Comment by carmocca** [comment]
Part of the motivation is also that a package like `lightning-transformers` can also use these abstractions without depending on `Flash`.

Would we put these in specific directory? Any name ideas?

**Comment by awaelchli** [comment]
would it bring dependencies? if so which ones and would they be optional or not?

**Comment by ethanwharris** [comment]
Thanks for comments :smiley: Some responses:

> Would we put these in specific directory? Any name ideas?

Yes, I think a seperate directory would be best, and to keep it as a sort of beta feature away from core. I guess we could call it `data_pipeline` or something like that.

> would it bring dependencies? if so which ones and would they be optional or not?

Shouldn't do, I think these classes only require stuff that we already have as required dependencies. If it did add any new dependencies, I think we would have them as optional since most users won't be using this.
